lib/active_model/serializer/association.rb

```diff
@@ -10,24 +10,86 @@ class Serializer
+        @parent_serializer = association_options.fetch(:parent_serializer)
+        @parent_serializer_options = association_options.fetch(:parent_serializer_options)
+        @include_slice = association_options.fetch(:include_slice)
+        @reflection = association_options.fetch(:reflection)
+
+        self.options = reflection.options.dup
+        # Pass the parent's namespace onto the child serializer
+        options[:namespace] ||= parent_serializer_options[:namespace]
+      end
+
       # @return [Symbol]
       def key
         options.fetch(:key, name)
       end
 
-      # @return [ActiveModel::Serializer, nil]
-      def serializer
-        options[:serializer]
-      end
-
       # @return [Hash]
       def links
         options.fetch(:links) || {}
       end
 
       # @return [Hash, nil]
       def meta
-        options[:meta]
+        reflection.options[:association_meta]
+      end
+
+      def object
+        @object ||= reflection.value(parent_serializer, include_slice)
+      end
+      alias eval_reflection_block object
+
+      def include_data?
+        eval_reflection_block if reflection.block
+        reflection.include_data?(include_slice)
+      end
+
+      # @return [ActiveModel::Serializer, nil]
+      def serializer
+        return @serializer if defined?(@serializer)
+        if serializer_class
+          serialize_object!(object)
+        elsif !object.nil? && !object.instance_of?(Object)
+          options[:virtual_value] = object
+        end
+        @serializer = options[:serializer]
+      end
+
+      def serialize_object!(object)
+        serializer = instantiate_serializer(object)
+        if serializer.nil?
+          # BUG: per #2027, JSON API resource relationships are only id and type, and hence either
+          # *require* a serializer or we need to be a little clever about figuring out the id/type.
+          # In either case, returning the raw virtual value will almost always be incorrect.
+          options[:virtual_value] = object.try(:as_json) || object
+        else
+          options[:serializer] = serializer
+        end
+      end
+
+      def instantiate_serializer(object)
+        serializer = catch(:no_serializer) do
+          serializer = options.fetch(:serializer, nil)
+          serializer_options = parent_serializer_options.except(:serializer)
+          serializer_options[:serializer] = serializer if serializer
+          serializer_options[:serializer_context_class] = parent_serializer.class
+          serializer_class.new(object, serializer_options)
+        end
+        serializer
+      end
+
+      def serializer_class
+        return @serializer_class if defined?(@serializer_class)
+        serializer_for_options = { namespace: options[:namespace] }
+        serializer_for_options[:serializer] = options[:serializer] if options.key?(:serializer)
+        @serializer_class = parent_serializer.class.serializer_for(object, serializer_for_options)
```

Comment:
1. This class is doing a lot more than it should. It has business logic (key/meta/links) and it also looks like a factory (`serializer`/`serialize_object`/`instantiate_serializer`/`serializer_class`). It's hard to maintain classes that you can understand what it's really meant to be doing, so it ends up having all sorts of methods. Perhaps we could replace all these methods with a class called... `Serializer`. See how association is doing the job a serializer again?
2. I've seen code like this in many other places. Perhaps we should just have it all in one place: `Serializer`. We already have a class called `Serializer`, I know, and that is doing things that are not responsibility of a serializer.